README.md
# S3 Object Streams

A small Node.js package that can be helpful when performing operations on very
large S3 buckets, those containing millions of objects or more, or when
processing S3 Inventory listings for those buckets. Streaming the listed
contents keeps memory under control and using the Streams API allows for fairly
compact utility code.

For very large buckets, S3 Inventory is always a better choice.

  * [S3ListObjectStream](#s3listobjectstream)
  * [S3ConcurrentListObjectStream](#s3concurrentlistobjectstream)
  * [S3UsageStream](#s3usagestream)
  * [S3InventoryUsageStream](#s3inventoryusagestream)

## Installing

Obtain via NPM:

```
npm install s3-object-streams
```

## S3ListObjectStream

An object stream that pipes in configuration objects for listing the contents of
an S3 bucket, and pipes out S3 object definitions.

```js
var AWS = require('aws-sdk');
var s3ObjectStreams = require('s3-object-streams');

var s3ListObjectStream = new s3ObjectStreams.S3ListObjectStream();
var s3Client = new AWS.S3();

// Log all of the listed objects.
s3ListObjectStream.on('data', function (s3Object) {
  console.info(s3Object);
});
s3ListObjectStream.on('end', function () {
  console.info('Listing complete.')
});
s3ListObjectStream.on('error', function (error) {
  console.error(error);
});

// List the contents of a couple of different buckets.
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket1',
  // Optional, only list keys with the given prefix.
  prefix: 'examplePrefix/',
  // Optional, group keys using a delimiter.
  // delimiter: '/',
  // Optional, defaults to 1000. The number of objects per request.
  maxKeys: 1000
});
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket2'
});
s3ListObjectStream.end();
```

Objects emitted by the stream have the standard format, with the addition of a
`Bucket` property:

```js
{
  Bucket: 'exampleBucket1',
  Key: 'examplePrefix/file.txt',
  LastModified: Date.now(),
  ETag: 'tag string',
  Size: 200,
  StorageClass: 'STANDARD',
  Owner: {
    DisplayName: 'exampleowner',
    ID: 'owner ID'
  }
```

## S3ConcurrentListObjectStream

This works in the same way as the `S3ListObjectStream`, but under the hood it
splits up the bucket by common prefixes and then recursively lists objects under
each common prefix concurrently, up to the maximum specified concurrency.

```js
var AWS = require('aws-sdk');
var s3ObjectStreams = require('s3-object-streams');

var s3ConcurrentListObjectStream = new s3ObjectStreams.S3ConcurrentListObjectStream({
  // Optional, defaults to 15.
  maxConcurrency: 15
});
var s3Client = new AWS.S3();

// Log all of the listed objects.
s3ConcurrentListObjectStream.on('data', function (s3Object) {
  console.info(s3Object);
});
s3ConcurrentListObjectStream.on('end', function () {
  console.info('Listing complete.')
});
s3ConcurrentListObjectStream.on('error', function (error) {
  console.error(error);
});

// List the contents of a couple of different buckets.
s3ConcurrentListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket1',
  // Optional, only list keys with the given prefix.
  prefix: 'examplePrefix/',
  // Optional, group keys using a delimiter.
  // delimiter: '/',
  // Optional, defaults to 1000. The number of objects per request.
  maxKeys: 1000
});
s3ConcurrentListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket2'
});
s3ConcurrentListObjectStream.end();
```

Objects emitted by the stream have the standard format, with the addition of a
`Bucket` property:

```js
{
  Bucket: 'exampleBucket1',
  Key: 'examplePrefix/file.txt',
  LastModified: Date.now(),
  ETag: 'tag string',
  Size: 200,
  StorageClass: 'STANDARD',
  Owner: {
    DisplayName: 'exampleowner',
    ID: 'owner ID'
  }
```

## S3UsageStream

A stream for keeping a running total of count and size of listed S3 objects by
bucket and key prefix. Useful for applications with a UI that needs to track
progress.

```js
var AWS = require('aws-sdk');
var s3ObjectStreams = require('s3-object-streams');

var s3ListObjectStream = new s3ObjectStreams.S3ListObjectStream();
var s3UsageStream = new s3ObjectStreams.S3UsageStream({
  // Determine folders from keys with this delimiter.
  delimiter: '/',
  // Group one level deep into the folders.
  depth: 1,
  // Only send a running total once every 100 objects.
  outputFactor: 100
});
var s3Client = new AWS.S3();

s3ListObjectStream.pipe(s3UsageStream);

var runningTotals;

// Log all of the listed objects.
s3UsageStream.on('data', function (totals) {
  runningTotals = totals;
  console.info(runningTotals);
});
s3UsageStream.on('end', function () {
  console.info('Final total: ', runningTotals);
});
s3UsageStream.on('error', function (error) {
  console.error(error);
});

// Obtain the total usage for these two buckets.
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket1'
});
s3ListObjectStream.write({
  s3Client: s3Client,
  bucket: 'exampleBucket2'
});
s3ListObjectStream.end();
```

The running total objects emitted by the stream have the following format:

```js
{
  path: 'exampleBucket/folder1',
  storageClass: {
    STANDARD: {
      // The number of files of this storage class.
      count: 55,
      // Total size in bytes of files in this storage class.
      size: 1232983
    },
    STANDARD_IA: {
      count: 0,
      size: 0
    },
    REDUCED_REDUNDANCY: {
      count: 2,
      size: 5638
    },
    GLACIER: {
      count: 0,
      size: 0
    }
  }
}
```

## S3InventoryUsageStream

A stream for keeping a running total of count and size of S3 objects by bucket
and key prefix, accepting objects from an S3 Inventory CSV file rather than
from the `listObjects` API endpoint.

```js
// Core.
var fs = require('fs');
var zlib = require('zlib');

// NPM.
var csv = require('csv');
var _ = require('lodash');
var s3ObjectStreams = require('s3-object-streams');

// Assuming that we already have the manifest JSON and a gzipped CSV data file
// downloaded from S3:
var manifest = require('/path/to/manifest.json');
var readStream = fs.createReadStream('/path/to/data.csv.gz');

var s3InventoryUsageStream = new s3ObjectStreams.S3InventoryUsageStream({
  // Determine folders from keys with this delimiter.
  delimiter: '/',
  // Group one level deep into the folders.
  depth: 1,
  // Only send a running total once every 100 objects.
  outputFactor: 100
});

var runningTotals;

// Log all of the listed objects.
s3UsageStream.on('data', function (totals) {
  runningTotals = totals;
  console.info(JSON.stringify(runningTotals, null, '  '));
});

var complete = _.once(function (error) {
  if (error) {
    console.error(error);
  }
  else {
    console.info('Complete');
  }
});

var gunzip = zlib.createGunzip();
var csvParser = csv.parse({
  // The manifest file defines the columns for the CSV. Specifying them here
  // ensures that the parser returns objects with properties named for the
  // columns.
  columns: manifest.fileSchema.split(', ')
});

csvParser.on('error', complete);
gunzip.on('error', complete);
readStream.on('error', complete);
s3InventoryUsageStream.on('error', complete);
s3InventoryUsageStream.on('end', complete);

// Unzip the file on the fly, feed it to the csvParser, and then into the
// object stream.
readStream.pipe(gunzip).pipe(csvParser).pipe(transformer).pipe(s3InventoryUsageStream);
```

The running total objects emitted by the stream have the following format:

```js
{
  path: 'exampleBucket/folder1',
  storageClass: {
    STANDARD: {
      // The number of files of this storage class.
      count: 55,
      // Total size in bytes of files in this storage class.
      size: 1232983
    },
    STANDARD_IA: {
      count: 0,
      size: 0
    },
    REDUCED_REDUNDANCY: {
      count: 2,
      size: 5638
    },
    GLACIER: {
      count: 0,
      size: 0
    }
  }
}
```

## Improvements
- [ ] upgrade async to v3.0.0
- [ ] refactor to use async/await


CHANGELOG.md
# Changelog

## 0.8.1

  * Add `maxdepth` option to S3ConcurrentObjectStream class
  * Move `aws-sdk` to peerDependencies
  * Add `npm test` script to make running tests easier
  * Push each `CommonPrefix` into the stream

## 0.8.0

  * Add delimiter option to S3ListObjectStream class.

## 0.7.0

  * Add S3InventoryUsage class.

## 0.6.2

  * Fix possible problem resulting from mistaken assumption about API behavior.

## 0.6.1

  * Add missing STANDARD_IA storage class.

## 0.6.0

  * A better algorithm for concurrent listing.

## 0.5.0

  * Split out usage counts by storage class.

## 0.4.2

  * Further fixing of the S3ConcurrentListObjectStream.

## 0.4.1

  * Fix issue with passing in `prefix/` rather than `prefix`.

## 0.4.0

  * Add retry on error to S3 API operations.

## 0.3.0

  * Add the S3ConcurrentListObjectStream class.

## 0.2.0

  * Add an outputFactor configuration parameter to S3UsageStream.

## 0.1.1

  * Fix bug in S3UsageStream.

## 0.1.0

  * Initial release.


test/lib/stream/s3ConcurrentListObjectStream.spec.js
/**
 * @fileOverview Tests for lib/stream/s3ConcurrentListObjectStream.
 */

// NPM.
var AWS = require('aws-sdk');
var _ = require('lodash');

// Local.
var S3ConcurrentListObjectStream = require('../../../lib/stream/s3ConcurrentListObjectStream');

describe('lib/stream/s3ConcurrentListObjectStream', function () {
  var listObjectsResponse1;
  var listObjectsResponse2;
  var prefix;
  var sandbox;
  var s3Client;
  var s3ConcurrentListObjectStream;

  beforeEach(function () {
    sandbox = sinon.sandbox.create();

    s3Client = new AWS.S3();
    prefix = 'prefix/';

    listObjectsResponse1 = {
      IsTruncated: true,
      NextContinuationToken: 'token',
      CommonPrefixes: [
        { Prefix: prefix + 'b/' },
        { Prefix: prefix + 'c/' }
      ],
      Contents: [
        {
          Key: prefix + 'a1'
        },
        {
          Key: prefix + 'a2'
        }
      ]
    };
    listObjectsResponse2 = {
      IsTruncated: false,
      // The prefixes listed already won't appear again.
      CommonPrefixes: [],
      Contents: [
        {
          Key: prefix + 'a3'
        }
      ]
    };

    sandbox.stub(s3Client, 'listObjectsV2');
    s3Client.listObjectsV2.onCall(0).yields(null, listObjectsResponse1);
    s3Client.listObjectsV2.onCall(1).yields(null, listObjectsResponse2);

    s3ConcurrentListObjectStream = new S3ConcurrentListObjectStream();
    sandbox.stub(s3ConcurrentListObjectStream, 'push');
  });

  afterEach(function () {
    sandbox.restore();
  });

  describe('listDirectoryPage', function () {
    var options;

    beforeEach(function () {
      options = {
        s3Client: s3Client,
        bucket: 'bucket',
        delimiter: '/',
        prefix: 'prefix/',
        continuationToken: 'token',
        maxKeys: 50
      };
    });

    it('functions as expected', function (done) {
      s3ConcurrentListObjectStream.listDirectoryPage(
        options,
        function (error, nextContinuationToken, s3Objects, commonPrefixes) {
          sinon.assert.calledWith(
            s3Client.listObjectsV2,
            {
              Bucket: options.bucket,
              Delimiter: options.delimiter,
              ContinuationToken: options.continuationToken,
              MaxKeys: options.maxKeys,
              Prefix: options.prefix
            },
            sinon.match.func
          );

          expect(nextContinuationToken).to.eql(listObjectsResponse1.NextContinuationToken);
          expect(commonPrefixes).to.eql(listObjectsResponse1.CommonPrefixes);
          expect(s3Objects).to.eql(listObjectsResponse1.Contents);

          done(error);
        }
      );
    });

    it('yields error on API error', function (done) {
      s3Client.listObjectsV2.onCall(0).yields(new Error());
      s3Client.listObjectsV2.onCall(1).yields(new Error());
      s3Client.listObjectsV2.onCall(2).yields(new Error());

      s3ConcurrentListObjectStream.listDirectoryPage(
        options,
        function (error, nextContinuationToken, commonPrefixes) {
          sinon.assert.callCount(s3Client.listObjectsV2, 3);
          expect(error).to.be.instanceOf(Error);
          done();
        }
      );
    });
  });

  describe('listDirectoryAndRecurse', function () {
    var options;

    beforeEach(function () {
      options = {
        s3Client: s3Client,
        bucket: 'bucket',
        delimiter: '/',
        prefix: prefix,
        maxKeys: 50
      };

      sandbox.stub(s3ConcurrentListObjectStream.queue, 'push');
    });

    it('functions as expected', function (done) {
      s3ConcurrentListObjectStream.listDirectoryAndRecurse(
        options,
        function (error) {
          sinon.assert.callCount(s3Client.listObjectsV2, 2);

          expect(s3Client.listObjectsV2.getCall(0).args[0]).to.eql({
            Bucket: options.bucket,
            Delimiter: options.delimiter,
            ContinuationToken: undefined,
            MaxKeys: options.maxKeys,
            Prefix: options.prefix
          });
          expect(s3Client.listObjectsV2.getCall(1).args[0]).to.eql({
            Bucket: options.bucket,
            Delimiter: options.delimiter,
            ContinuationToken: listObjectsResponse1.NextContinuationToken,
            MaxKeys: options.maxKeys,
            Prefix: options.prefix
          });

          sinon.assert.callCount(s3ConcurrentListObjectStream.push, 5);

          expect(s3ConcurrentListObjectStream.push.getCall(0).args).to.eql([
            listObjectsResponse1.CommonPrefixes[0]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(1).args).to.eql([
            listObjectsResponse1.CommonPrefixes[1]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(2).args).to.eql([
            listObjectsResponse1.Contents[0]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(3).args).to.eql([
            listObjectsResponse1.Contents[1]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(4).args).to.eql([
            listObjectsResponse2.Contents[0]
          ]);

          sinon.assert.callCount(s3ConcurrentListObjectStream.queue.push, 2);

          expect(s3ConcurrentListObjectStream.queue.push.getCall(0).args[0]).to.eql({
            s3Client: s3Client,
            bucket: 'bucket',
            delimiter: '/',
            prefix: prefix + 'b/',
            maxKeys: 50,
            currentdepth: 2
          });
          expect(s3ConcurrentListObjectStream.queue.push.getCall(1).args[0]).to.eql({
            s3Client: s3Client,
            bucket: 'bucket',
            delimiter: '/',
            prefix: prefix + 'c/',
            maxKeys: 50,
            currentdepth: 2
          });

          done(error);
        }
      );
    });

    it('retries and yields error on API error', function (done) {
      s3Client.listObjectsV2.onCall(0).yields(new Error());
      s3Client.listObjectsV2.onCall(1).yields(new Error());
      s3Client.listObjectsV2.onCall(2).yields(new Error());

      s3ConcurrentListObjectStream.listDirectoryAndRecurse(
        options,
        function (error, commonPrefixes) {
          sinon.assert.callCount(s3Client.listObjectsV2, 3);
          expect(error).to.be.instanceOf(Error);
          done();
        }
      );
    });

    it('stops when maxdepth is reached', function (done) {
      s3ConcurrentListObjectStream.options.maxdepth = 1
      s3ConcurrentListObjectStream.listDirectoryAndRecurse(
        options,
        function (error) {
          sinon.assert.callCount(s3Client.listObjectsV2, 2);

          expect(s3Client.listObjectsV2.getCall(0).args[0]).to.eql({
            Bucket: options.bucket,
            Delimiter: options.delimiter,
            ContinuationToken: undefined,
            MaxKeys: options.maxKeys,
            Prefix: options.prefix
          });
          expect(s3Client.listObjectsV2.getCall(1).args[0]).to.eql({
            Bucket: options.bucket,
            Delimiter: options.delimiter,
            ContinuationToken: listObjectsResponse1.NextContinuationToken,
            MaxKeys: options.maxKeys,
            Prefix: options.prefix
          });

          sinon.assert.callCount(s3ConcurrentListObjectStream.push, 5);

          expect(s3ConcurrentListObjectStream.push.getCall(0).args).to.eql([
            listObjectsResponse1.CommonPrefixes[0]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(1).args).to.eql([
            listObjectsResponse1.CommonPrefixes[1]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(2).args).to.eql([
            listObjectsResponse1.Contents[0]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(3).args).to.eql([
            listObjectsResponse1.Contents[1]
          ]);
          expect(s3ConcurrentListObjectStream.push.getCall(4).args).to.eql([
            listObjectsResponse2.Contents[0]
          ]);

          sinon.assert.callCount(s3ConcurrentListObjectStream.queue.push, 0);

          done(error);
        }
      );
    })
  });

  describe('processIncomingObject', function () {
    var options;

    beforeEach(function () {
      options = {
        s3Client: s3Client,
        bucket: 'bucket',
        prefix: prefix
      };

      sandbox.stub(s3ConcurrentListObjectStream.queue, 'push');
    });

    it('functions as expected', function (done) {
      s3ConcurrentListObjectStream.processIncomingObject(options, function (error) {
        done(error);
      });

      sinon.assert.calledWith(
        s3ConcurrentListObjectStream.queue.push,
        sinon.match.object,
        sinon.match.func
      );

      var argOpts = s3ConcurrentListObjectStream.queue.push.getCall(0).args[0];
      expect(argOpts.s3Client).to.equal(options.s3Client);
      expect(argOpts.bucket).to.equal(options.bucket);
      expect(argOpts.delimiter).to.equal('/');
      expect(argOpts.prefix).to.equal(options.prefix);

      s3ConcurrentListObjectStream.queue.drain()
    });

    it('yields error for missing options', function (done) {
      s3ConcurrentListObjectStream.processIncomingObject(null, function (error) {
        expect(error).to.be.instanceOf(Error);
        done();
      });
    });

    it('yields error for missing options.s3Client', function (done) {
      delete options.s3Client;

      s3ConcurrentListObjectStream.processIncomingObject(options, function (error) {
        expect(error).to.be.instanceOf(Error);
        done();
      });
    });

    it('yields error for missing options.bucket', function (done) {
      delete options.bucket;

      s3ConcurrentListObjectStream.processIncomingObject(options, function (error) {
        expect(error).to.be.instanceOf(Error);
        done();
      });
    });
  });
});


lib/stream/s3ConcurrentListObjectStream.js
/**
 * @fileOverview S3ConcurrentListObjectStream class definition.
 */

// Core.
var Transform = require('stream').Transform;
var util = require('util');

// NPM.
var async = require('async');
var _ = require('lodash');

//---------------------------------------------------------------------------
// Class constructor.
//---------------------------------------------------------------------------

/**
 * @class An object stream to list S3 objects via multiple concurrent requests.
 *
 * This is considerably faster than S3ListObjectStream, but the level of
 * concurrency is limited by practical concerns. Aside from API throttling,
 * most ordinary machines start to exhibit failures with AWS SDKs at around 20
 * parallel requests.
 *
 * It works by descending through 'directories' defined by splitting keys with
 * the delimiter. For each new 'subdirectory', a new task is spawned to list
 * it and its 'subdirectories'. Those tasks run in parallel up to the maximum
 * concurrency.
 *
 * Pipe in objects of the following form:
 *
 * {
 *   s3Client: new AWS.S3(),
 *   bucket: 'exampleBucket',
 *   // Optional. Maximum degree of concurrency if there are many common
 *   // prefixes. Defaults to 10.
 *   maxConcurrency: 10,
 *   // Optional. Used to find common prefixes that can be listed concurrently.
 *   // Defaults to '/'.
 *   delimiter: '/',
 *   // Optional, defaults to 1000. How many objects to return in one API
 *   // request under the hood.
 *   maxKeys: 1000,
 *   // Optional. If present, only list objects with keys matching the prefix.
 *   prefix: 'examplePrefix'
 * }
 *
 * Pipe out standard response objects from the S3 listObjects API, with the
 * addition of the bucket name as well.
 *
 * {
 *   Bucket: 'exampleBucket',
 *   Key: ...
 *   LastModified: ...
 *   ETag: ...
 *   Size: ...
 *   StorageClass: ...
 *   Owner: {
 *     DisplayName: ...
 *     ID: ...
 *   }
 * }
 *
 * @param {Object} options Standard stream options, plus the following.
 * @param {Number} [options.maxConcurrency] Defaults to 15. Number of
 *   concurrent API requests to make.
 * @param {Number} [options.maxdepth] Default to -1. How far down the rabbit hole
 *   do we go
 */
function S3ConcurrentListObjectStream (options) {
  options = options || {};
  // Important; make this an object stream.
  options.objectMode = true;
  options.maxConcurrency = options.maxConcurrency || 15;
  options.maxdepth = options.maxdepth || -1;
  this.options = options

  // A queue for managing concurrency of API requests.
  this.queue = async.queue(
    this.listDirectoryAndRecurse.bind(this),
    options.maxConcurrency
  );

  S3ConcurrentListObjectStream.super_.call(this, options);
}

util.inherits(S3ConcurrentListObjectStream, Transform);

//---------------------------------------------------------------------------
// Methods
//---------------------------------------------------------------------------

/**
 * Helper function. Send a task to the queue.
 *
 * @param {Object} options
 * @param {AWS.S3} options.s3Client An AWS client instance.
 * @param {String} options.bucket The bucket to list.
 * @param {Number} options.delimiter Used to find common prefixes to split out
 *   requests for object listing. Defaults to '/'.
 * @param {Number} [options.maxKeys] Maximum number of keys to return per
 *   request. Defaults to 1000.
 * @param {String} [options.prefix] If present, only list objects with keys that
 *   match the prefix.
 */
S3ConcurrentListObjectStream.prototype.sendToQueue = function (options) {
  var self = this;

  this.queue.push(options, function (error) {
    if (error) {
      self.emit('error', error);
    }
  });
};

/**
 * List the objects in a given 'directory' by common prefix, and all the
 * common prefixes for 'subdirectories'.
 *
 * @param {Object} options
 * @param {AWS.S3} options.s3Client An AWS client instance.
 * @param {String} options.bucket The bucket name.
 * @param {Number} options.delimiter Used to find common prefixes to split out
 *   requests for object listing. Defaults to '/'.
 * @param {String} [options.prefix] If set only return keys beginning with
 *   the prefix value.
 * @param {String} [options.continuationToken] If set the list only a paged set
 *   of keys, with the token showing the start point.
 * @param {Number} [options.maxKeys] Maximum number of keys to return per
 *   request. Defaults to 1000.
 * @param {Function} callback - Callback of the form
    function (error, nextMarker, Object[], String[]).
 */
S3ConcurrentListObjectStream.prototype.listDirectoryPage = function (
  options,
  callback
) {
  var params = {
    Bucket: options.bucket,
    Delimiter: options.delimiter,
    ContinuationToken: options.continuationToken,
    MaxKeys: options.maxKeys,
    Prefix: options.prefix
  };

  // S3 operations have a small but significant error rate.
  async.retry(
    3,
    function (asyncCallback) {
      options.s3Client.listObjectsV2(params, asyncCallback);
    },
    function (error, response) {
      var continuationToken;

      if (error) {
        return callback(error);
      }

      // Check to see if there are yet more objects to be obtained, and if so
      // return the continuationToken for use in the next request.
      if (response.IsTruncated) {
        continuationToken = response.NextContinuationToken;
      }

      callback(
        null,
        continuationToken,
        response.Contents,
        response.CommonPrefixes
        // _.map(response.CommonPrefixes, function (prefixObject) {
        //   return prefixObject.Prefix;
        // })
      );
    }
  );
};

/**
 * List the objects in a given 'directory' by common prefix, and spawn new tasks
 * to list all child 'directories'.
 *
 * All the objects found in this 'directory' are piped onward.
 *
 * @param {Object} options
 * @param {AWS.S3} options.s3Client An AWS client instance.
 * @param {String} options.bucket The bucket to list.
 * @param {Number} options.delimiter Used to find common prefixes to split out
 *   requests for object listing. Defaults to '/'.
 * @param {Number} [options.maxKeys] Maximum number of keys to return per
 *   request. Defaults to 1000.
 * @param {String} [options.prefix] If present, only list objects with keys that
 *   match the prefix.
 * @param {Function} callback Of the form function (error).
 */
S3ConcurrentListObjectStream.prototype.listDirectoryAndRecurse = function (
  options,
  callback
) {
  var self = this;
  const { maxdepth } = this.options

  /**
   * Recursively list common prefixes.
   *
   * @param {String|undefined} marker A value provided by the S3 API to enable
   *   paging of large lists of keys. The result set requested starts from the
   *   marker. If not provided, then the list starts from the first key.
   */
  function listRecursively (continuationToken) {
    var currentdepth = options.currentdepth || 1
    options.continuationToken = continuationToken;

    self.listDirectoryPage(
      options,
      function (error, nextContinuationToken, s3Objects, commonPrefixes) {
        if (error) {
          return callback(error);
        }

        // Each common prefix is only returned once, even across requests using
        // continuation tokens for 'subdirectories' containing many files.
        _.each(commonPrefixes, function (commonPrefix) {
          self.push(commonPrefix);
          // only add to queue if we haven't reached maxdepth yet
          if (maxdepth === -1 || maxdepth > currentdepth) {
            var subDirectoryOptions = _.chain(
              {}
            ).extend(
              options,
              {
                prefix: commonPrefix.Prefix,
                currentdepth: currentdepth + 1
              }
            ).omit(
              'continuationToken'
            ).value();

            self.sendToQueue(subDirectoryOptions);
          }
        });

        // Send any S3 object definitions to be piped onwards.
        s3Objects.forEach(function (object) {
          object.Bucket = options.bucket;
          self.push(object);
        });

        // If there are more objects, go get them.
        if (nextContinuationToken) {
          listRecursively(nextContinuationToken);
        }
        else {
          callback();
        }
      }
    );
  }

  // Start the recursive listing at the beginning, with no continuationToken.
  listRecursively();
};

/**
 * List objects from S3 and push them to the stream, according to the details
 * in the incoming object.
 *
 * @param {Object} options
 * @param {AWS.S3} options.s3Client An AWS client instance.
 * @param {String} options.bucket The bucket to list.
 * @param {Number} [options.maxConcurrency] The limit on concurrent API
 *   requests. Defaults to 10.
 * @param {Number} [options.delimiter] Used to find common prefixes to split out
 *   requests for object listing. Defaults to '/'.
 * @param {Number} [options.maxKeys] Maximum number of keys to return per
 *   request. Defaults to 1000.
 * @param {String} [options.prefix] If present, only list objects with keys that
 *   match the prefix.
 * @param {String} encoding Irrelevant since this is an object stream.
 * @param {Function} callback Invoked after this listing is processed.
 */
S3ConcurrentListObjectStream.prototype.processIncomingObject = function (
  options,
  callback
) {
  if (!options || typeof options !== 'object') {
    return callback(new Error('An object is expected.'));
  }
  if (!options.s3Client) {
    return callback(new Error('Missing options.s3Client'));
  }
  if (!options.bucket) {
    return callback(new Error('Missing options.bucket'));
  }

  options.delimiter = options.delimiter || '/';

  // Reset the global state for the next item.
  //
  // Since incoming items are processed in order, the next not starting until
  // the callback for the first invoked, it is safe to manipulate the global
  // state in this way.
  this.commonPrefixes = {};
  this.queue.drain = _.once(callback);

  // Start things going.
  this.sendToQueue(options);
};

//---------------------------------------------------------------------------
// Overridden methods.
//---------------------------------------------------------------------------

/**
 * Implementation of the necessary transform method.
 *
 * @param {Object} data A listObjects configuration object since this is an
 *   object stream.
 * @param {String} encoding Irrelevant since this is an object stream.
 * @param {Function} callback Invoked after this listing is processed.
 */
S3ConcurrentListObjectStream.prototype._transform = function (
  data,
  encoding,
  callback
) {
  this.processIncomingObject(data, callback);
};

//---------------------------------------------------------------------------
// Export class constructor.
//---------------------------------------------------------------------------

module.exports = S3ConcurrentListObjectStream;
